### PR TITLE
URL form decoder tweaks

### PR DESCRIPTION
**Pull request by stephencelis**
Just making the strategies we use a bit more honest and documenting them along the way. Wrote these while waiting for food and just committed em now. Will probably merge so I can pin `PointFree` to the latest on my branch!

### CODE REVIEW

File: Sources/UrlFormEncoding/UrlFormDecoder.swift [hunk 1]
```diff
@@ -10,21 +10,37 @@ public final class UrlFormDecoder: Decoder {
   public private(set) var codingPath: [CodingKey] = []
   public var dataDecodingStrategy: DataDecodingStrategy = .deferredToData
   public var dateDecodingStrategy: DateDecodingStrategy = .deferredToDate
-  public var parsingStrategy: ParsingStrategy = .accumulatePairs
+  public var parsingStrategy: ParsingStrategy = .accumulateValues
```
Comment (stephencelis) [inline Sources/UrlFormEncoding/UrlFormDecoder.swift]: Don't really know why I named it the other way. Was dealing with a lot of pairs at the time.

File: Sources/UrlFormEncoding/UrlFormDecoder.swift [hunk 2]
```diff
@@ -10,21 +10,37 @@ public final class UrlFormDecoder: Decoder {
...
 
   public init() {
   }
 
   public func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
-    let container = self.parsingStrategy.strategy(String(decoding: data, as: UTF8.self))
+    let query = String(decoding: data, as: UTF8.self)
+    let container: [String: Any]
+    switch self.parsingStrategy {
+    case .accumulateValues:
```
Comment (stephencelis) [inline Sources/UrlFormEncoding/UrlFormDecoder.swift]: Kinda funny to get here after that struct vs. enum convo last week! The one interesting thing that ya lose when moving from an `enum` to a `struct` that wraps a closure is the ability to tag and switch on the cases!

File: Sources/UrlFormEncoding/UrlFormDecoder.swift [hunk 3]
```diff
@@ -10,21 +10,37 @@ public final class UrlFormDecoder: Decoder {
...
+    }
     self.containers.append(container)
     defer { self.containers.removeLast() }
     return try T(from: self)
   }
 
+  private func unbox(_ value: Any) -> String? {
+    switch self.parsingStrategy {
+    case .accumulateValues:
+      return value as? String ?? (value as? [String])?.last
```
Comment (stephencelis) [inline Sources/UrlFormEncoding/UrlFormDecoder.swift]: We now only auto-unbox array values for the default, flat way.

**Comment by stephencelis** [comment]
Wow, looks like Linux and macOS have different behavior for `Array.removeLast`! Gonna file a bug.

**Comment by stephencelis** [comment]
On closer inspection, the behavior's the same! Not sure what is preventing the crash on macOS, but filed a bug here: https://bugs.swift.org/browse/SR-6598

File: Sources/UrlFormEncoding/UrlFormDecoder.swift [hunk 4]
```diff
@@ -510,24 +526,71 @@ public final class UrlFormDecoder: Decoder {
...
-      for (name, value) in pairs(query) {
-        var values = params[name] as? [Any] ?? []
-        values.append(value)
-        params[name] = values
-      }
-      return params
-    }
-
+  public enum ParsingStrategy {
+    /// A parsing strategy that accumulates values when multiple keys are provided.
```
Comment (mbrandonw) [inline Sources/UrlFormEncoding/UrlFormDecoder.swift]: amazing docs! we need to do this more after launch and use jazzy or something to generate docs

**Pull request merged**